lib/Transforms/SimplifyParsers.cpp

```diff
@@ -0,0 +1,114 @@
+#include <llvm/Support/ErrorHandling.h>
+
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "p4mlir/Dialect/P4HIR/ParserGraph.h"
+#include "p4mlir/Transforms/Passes.h"
+
+#define DEBUG_TYPE "p4hir-simplify-parsers"
+
+using namespace mlir;
+
+namespace P4::P4MLIR {
+#define GEN_PASS_DEF_SIMPLIFYPARSERS
+#include "p4mlir/Transforms/Passes.cpp.inc"
+
+namespace {
+struct SimplifyParsers : public impl::SimplifyParsersBase<SimplifyParsers> {
+    void runOnOperation() override;
+
+ private:
+    /// Collapses linear sequences of states without branches or annotations.
+    void collapseChains(P4HIR::ParserOp parser);
+};
+}  // end anonymous namespace
+
+void SimplifyParsers::collapseChains(P4HIR::ParserOp parser) {
+    // TODO: Revisit this to use ParserCallGraph instead
+    mlir::DenseMap<P4HIR::ParserStateOp, unsigned> indegree;
+    mlir::DenseMap<P4HIR::ParserStateOp, unsigned> outdegree;
+    for (auto state : parser.states()) {
+        for (auto next : state.getNextStates()) {
+            ++outdegree[state];
+            ++indegree[next];
+        }
+    }
+
+    // succ[s1] = s2 if there is exactly one outgoing edge from s1 to s2.
+    mlir::DenseMap<P4HIR::ParserStateOp, P4HIR::ParserStateOp> succ;
+    // pred[s2] = s1 if there is exactly one outgoing edge from s1 to s2.
+    mlir::DenseMap<P4HIR::ParserStateOp, P4HIR::ParserStateOp> pred;
+
+    // We diconnect any annotated states since they can't be collapsed
+    // and if we kept them they'd poison any merging downstream.
+    for (auto state : parser.states()) {
+        if (outdegree[state] != 1 || state.getAnnotations()) continue;
+        P4HIR::ParserStateOp successor = *state.getNextStates().begin();
+        if (indegree[successor] != 1 || successor.getAnnotations()) continue;
+
+        succ[state] = successor;
+        pred[successor] = state;
```

Comment:
Can we simplify here a bit? We just need heads here, right? Can we just check if their outdegree is 1 (plus usual annotations checks) and indegree != 1. And make a worklist of heads? Then you'd just iterate below over the worklist.